provider/civo/civo.go

```diff
@@ -0,0 +1,537 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package civo
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/civo/civogo"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/pkg/apis/externaldns"
+	"sigs.k8s.io/external-dns/plan"
+	"sigs.k8s.io/external-dns/provider"
+)
+
+// CivoProvider is an implementation of Provider for Civo's DNS.
+type CivoProvider struct {
+	provider.BaseProvider
+	Client       civogo.Client
+	domainFilter endpoint.DomainFilter
+	DryRun       bool
+}
+
+// CivoChanges All API calls calculated from the plan
+type CivoChanges struct {
+	Creates []*CivoChangeCreate
+	Deletes []*CivoChangeDelete
+	Updates []*CivoChangeUpdate
+}
+
+// CivoChangeCreate Civo Domain Record Creates
+type CivoChangeCreate struct {
+	Domain  civogo.DNSDomain
+	Options *civogo.DNSRecordConfig
+}
+
+// CivoChangeUpdate Civo Domain Record Updates
+type CivoChangeUpdate struct {
+	Domain       civogo.DNSDomain
+	DomainRecord civogo.DNSRecord
+	Options      civogo.DNSRecordConfig
+}
+
+// CivoChangeDelete Civo Domain Record Deletes
+type CivoChangeDelete struct {
+	Domain       civogo.DNSDomain
+	DomainRecord civogo.DNSRecord
+}
+
+// NewCivoProvider initializes a new Civo DNS based Provider.
+func NewCivoProvider(domainFilter endpoint.DomainFilter, dryRun bool) (*CivoProvider, error) {
+	token, ok := os.LookupEnv("CIVO_TOKEN")
+	if !ok {
+		return nil, fmt.Errorf("no token found")
+	}
+
+	civoClient, err := civogo.NewClient(token, "LON1")
```

Comment:
`LON1` is a region right? If so, it shouldn't be hardcoded, but configurable instead.

Reply:
That region is just for the client, the DNS is not per region is something globally, I modify the code to explain that